src/sagas/task.saga.test.js

```diff
@@ -0,0 +1,393 @@
+import { call, put, select } from 'redux-saga/effects';
+import { cloneableGenerator } from 'redux-saga/utils';
+import * as childProcess from 'child_process';
+import * as path from 'path';
+import chalkRaw from 'chalk';
+
+import {
+  launchDevServer,
+  taskRun,
+  taskAbort,
+  displayTaskComplete,
+  taskComplete,
+  getDevServerCommand,
+} from './task.saga';
+import {
+  attachTaskMetadata,
+  receiveDataFromTaskExecution,
+  completeTask,
+  loadDependencyInfoFromDisk,
+} from '../actions';
+import killProcessId from '../services/kill-process-id.service';
+import { getProjectById } from '../reducers/projects.reducer';
+import { getPathForProjectId } from '../reducers/paths.reducer';
+import findAvailablePort from '../services/find-available-port.service';
+import {
+  getBaseProjectEnvironment,
+  PACKAGE_MANAGER_CMD,
+} from '../services/platform.service';
+import { ipcRenderer } from 'electron';
+
+const chalk = new chalkRaw.constructor({ level: 3 });
+
+jest.mock('electron', () => ({
+  ipcRenderer: {
+    send: jest.fn(),
+  },
+  remote: {
+    app: { getAppPath: () => '' },
+  },
+}));
+
+jest.mock('uuid/v1', () => () => 'mocked-uuid-v1');
+
+// we don't actually need to run these commands through
+// the terminal, so an object with similar shape to that
+// returned by `child_process.spawn` should be sufficient.
+// Mock the relevant methods so they can be spied on.
+const mockSpawn = processId => ({
+  pid: processId,
+  write: jest.fn(),
+  on: jest.fn(),
+  stdin: {
+    write: jest.fn(),
+    on: jest.fn(),
+  },
+  stdout: {
+    write: jest.fn(),
+    on: jest.fn(),
+  },
+  stderr: {
+    write: jest.fn(),
+    on: jest.fn(),
+  },
+});
+
+describe('task saga', () => {
+  beforeAll(() => {
+    // `getBaseProjectEnvironment` appends to the existing
+    // `window.process.env.PATH`, so make sure it's defined
+    if (!window.process) {
+      window.process = { env: { PATH: '' } };
+    }
+  });
+
+  describe('launchDevServer saga', () => {
+    it('should throw if no task is provided', () => {
+      expect(() => launchDevServer()).toThrow();
+    });
+
+    const task = { projectId: 'pickled-tulip' };
+    const saga = cloneableGenerator(launchDevServer)({ task });
+    const project = { type: 'create-react-app' };
+    const projectPath = '/path/to/project';
+    const port = 3000;
+    const { args, env } = getDevServerCommand(task, project.type, port);
+    const processId = 12345;
+    const mockProcess = mockSpawn(processId);
```

Comment:
Huh, `cloneableGenerator` is neat. I could have used that in my tests, will update at some point.

It's a bit weird to me to see random constants defined in a `describe`, I'm used to seeing this in a `beforeAll` block... but I suppose if it works, there's no problem!

Reply:
You'll have to excuse me, I'm still figuring out Jest conventions.

On that note, how does `beforeAll()` work with block-scoped variables? If I do

```js
beforeAll(() => {
  const saga = /* ... */;
});
```

can I then do

```js
it('should do whatever', () => {
  saga.next();
});
```

or do I first need to declare it in the parent scope and set its value in `beforeAll()`?

Reply:
Let me know and I'll rewrite these.

Reply:
Yeah, that's right:

```js
let thing;
beforeAll(() => {
  thing = 'yeah';
});
```

That said, I don't think it matters at all. No rewriting necessary :) I've seen it done that way but I don't actually know why (maybe for parallelization somehow?). The way you have it works :+1: